### ISSUE TITLE
ENH improve speed of fitting Cox model by relying on the fast skglm solver

### ISSUE THREAD

**Issue by mathurinm** [CamDavidsonPilon/lifelines#1531]
I am one of the developers of ``skglm``, a python package that improves ``scikit-learn`` for Generalized Linear Models by providing more functionalities and faster solvers. 
We have recently worked on a solver for the Cox estimator, for which ``lifelines`` provides a reference implementation. 

Preliminary results indicate time speedups up to x500 when using `skglm`

![Timing comparison between skglm and lifelines](https://github.com/CamDavidsonPilon/lifelines/assets/8993218/61c585b1-a00a-4994-bf7e-4aaff3da25e7)

Here is a [notebook](https://contrib.scikit-learn.org/skglm/auto_examples/plot_survival_analysis.html#sphx-glr-auto-examples-plot-survival-analysis-py) to illustrate the peformance and to showcase the ``scikit-learn`` like API of ``skglm``.  Also, here are the [results of a complete benchmark with Benchopt](https://badr-moufad.github.io/share-benchmarks/cox_bench/benchmark_cox_benchopt_run_2023-05-25_21h36m26.html) and the link to the [benchmark repo](https://github.com/benchopt/benchmark_l1_cox) to reproduce it.

In addition, some `skglm` features might be useful to the users of `lifelines`:
- support of design matrix with more columns than rows ([may cause issue in ``lifelines``](https://github.com/CamDavidsonPilon/lifelines/blob/master/lifelines/utils/__init__.py#L989))
- support of sparse design matrix (currently not supported in ``lifelines``)
- immediate extension to other penalizers such as Weighted L1, non convex regularizers, group Lasso penalty, etc

Based on this, we'd like to discuss the potential integration of ``skglm`` solver into ``lifelines`` for fitting the Cox Estimator.

A noteworthy point is that ``skglm`` relies heavily on ``numba`` JIT compilation, which may introduce a slight overhead during the initial model fit. However, this inconvenience is compensated by the gained advantages namely handling datasets with thousands of features and samples within a reasonable time.

We'd be happy to have your feedback on this.

Also pinging @Badr-Moufad @PABannier @QB3

**Comment by CamDavidsonPilon** [comment]
Wow that's very impressive! One thing I think you should try is to bin the times into buckets (as tied times are common in survival datasets, as we are often rounding to months, days, hours, etc.). The Cox model works by sorting times, but when there are ties, it has to use a technique to handle them. There are a few technique to handle ties: random, Efron, Breslow, exact (the most accurate, but slowest). Lifelines uses Efron's method, as its accuracy-to-speed tradeoff is good.

**Comment by mathurinm** [comment]
Indeed, we are working on adding support for Efron handling of ties here : https://github.com/scikit-learn-contrib/skglm/pull/159, it should be merged shortly.

**Comment by CamDavidsonPilon** [comment]
Very exciting work, team!

**Comment by Badr-MOUFAD** [comment]
Thank you again for your interest!
Here are the key improvement factors

- Levaraging the sparse nature of the solution with state-of-the-art working set strategy detailed in [our Neurips 2022 paper](https://proceedings.neurips.cc/paper_files/paper/2022/hash/fe5c31e525e9a26a1426ab0b589f42fe-Abstract-Conference.html) (Algo 1 and 2)
- Usage of Proximal Newton solver with diagonal upper-bound on the Hessian resulting in a linear computational and memory cost ([``skglm`` tutorial equation 6](https://contrib.scikit-learn.org/skglm/tutorials/cox_datafit.html#equation-diagonal-upper-bound))
- Efficient implementation of Cox datafit which achieves a linear cost of evaluating its value, gradient, and Hessian ([``skglm`` Cox implementation](https://github.com/scikit-learn-contrib/skglm/pull/168))

We are happy to discuss options for integrating ``skglm`` into ``lifelines``.
